src/Workspaces/Core/Portable/Workspace/Workspace.cs

```diff
@@ -707,14 +715,15 @@ protected internal void OnDocumentsAdded(ImmutableArray<DocumentInfo> documentIn
         {
             this.SetCurrentSolution(
                 static (oldSolution, data) => oldSolution.AddDocuments(data.documentInfos),
+                data: (@this: this, documentInfos),
                 onAfterUpdate: static (oldSolution, newSolution, data) =>
                 {
                     // Raise ProjectChanged as the event type here. DocumentAdded is presumed by many callers to have a
                     // DocumentId associated with it, and we don't want to be raising multiple events.
 
                     foreach (var projectId in data.documentInfos.Select(i => i.Id.ProjectId).Distinct())
                         data.@this.RaiseWorkspaceChangedEventAsync(WorkspaceChangeKind.ProjectChanged, oldSolution, newSolution, projectId);
-                }, (@this: this, documentInfos));
+                });
```

Comment:
complex case from last PR.  here we don't want to issue a single event, but many.  so we take full control and define the onAfterUpdate behavior ourselves.